File: Intermediate/15.Find the majority element in an array (element that appears more than n2 times)/Solution.py
def find_majority_element(arr):
    count = 0
    candidate = -1

    # Finding candidate
    for num in arr:
        if count == 0:
            candidate = num
            count = 1
        elif num == candidate:
            count += 1
        else:
            count -= 1

    # Verifying candidate
    count = 0
    for num in arr:
        if num == candidate:
            count += 1

    return candidate if count > len(arr) // 2 else -1

if __name__ == "__main__":
    n = int(input("Enter the size of the array: "))
    arr = list(map(int, input(f"Enter {n} elements of the array: ").split()))

    result = find_majority_element(arr)
    if result != -1:
        print(f"The majority element is: {result}")
    else:
        print("No majority element found.")

File: Intermediate/15.Find the majority element in an array (element that appears more than n2 times)/Solution.c
#include <stdio.h>

int findMajorityElement(int arr[], int n) {
    int count = 0, candidate = -1;

    // Finding candidate
    for (int i = 0; i < n; i++) {
        if (count == 0) {
            candidate = arr[i];
            count = 1;
        } else if (arr[i] == candidate) {
            count++;
        } else {
            count--;
        }
    }

    // Verifying candidate
    count = 0;
    for (int i = 0; i < n; i++) {
        if (arr[i] == candidate)
            count++;
    }
    return (count > n / 2) ? candidate : -1;
}

int main() {
    int n;
    printf("Enter the size of the array: ");
    scanf("%d", &n);

    int arr[n];
    printf("Enter %d elements of the array: ", n);
    for (int i = 0; i < n; i++) {
        scanf("%d", &arr[i]);
    }

    int result = findMajorityElement(arr, n);
    if (result != -1)
        printf("The majority element is: %d\n", result);
    else
        printf("No majority element found.\n");

    return 0;
}
